File: credit.c
#include <stdio.h>
#include <cs50.h>

//Ian Luria cs50

// Returns a specific digit based on multiples of ten
int getDigit(long tenCounter, long userCCNumber)
{
    int digit = (userCCNumber % tenCounter) /(tenCounter*0.1);
    return digit;
}

int main(void)
{
    long userCCNumber;
    
    do
    {
        userCCNumber = get_long("Your card number: ");
    } while (userCCNumber < 1000000000000);
    
    int grandSum = 0;
    int underlinedSum = 0;
    int notUnderlinedSum = 0;
    int counter = 1;
    long tenCounter = 10;
    
    while (tenCounter <= userCCNumber * 10)
    {   
        int digit = getDigit(tenCounter, userCCNumber);
        
        if (counter % 2 == 0)
        {
            int product = digit * 2;
            int underlined = product % 10;
            if (product > 9)
            {
                // Add the two digits together i.e. 16 = 7 
                underlined = ((product - underlined) * 0.1) + underlined;
            }
            
            underlinedSum += underlined;
        } 
        else
        {
            notUnderlinedSum += digit;
        }
        
        counter+=1;
        tenCounter *= 10;
    }
    
   
    counter -= 1;
    grandSum = underlinedSum + notUnderlinedSum;
    
    if (grandSum % 10 == 0)
    {
       char creditCard;
       int firstTwoDigits = (getDigit(tenCounter *0.1, userCCNumber)*10)+(getDigit(tenCounter*0.01, userCCNumber));
       int firstDigit = firstTwoDigits * 0.1; 
 
        if (firstTwoDigits == 34 || firstTwoDigits == 37)
        {
            if (counter == 15)
            {
                creditCard = 'A'; //AmEx
            } 
            else
            {
                creditCard = 'X'; //Fail
            }
        } 
        else if (firstDigit == 4)
        {  
            if (counter == 16 || counter == 13)
            {
                creditCard = 'V'; //Visa
            } 
            else
            {
                creditCard = 'X';
            }
        } 
        else if (counter == 16)
        {
            bool masterCard = false;
            
            for (int i = 51; i < 56; i++)
            {
                if (i == firstTwoDigits)
                {
                    masterCard = true;
                }    
            }
            
            if (masterCard)
            {
                creditCard = 'M'; //Mastercard
            } 
            else
            {
                creditCard = 'X';
            }
            
        } 
        else
        {
            creditCard = 'Z'; 
        }
   
        if (creditCard == 'Z')
        {
            printf("INVALID\n");
        } 
        else if (creditCard == 'V')
        {
            printf("VISA\n");
        } 
        else if (creditCard == 'M')
        {
            printf("MASTERCARD\n");
        } 
        else if (creditCard == 'A')
        {
            printf("AMEX\n");
        }
       
    } 
    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
from cs50 import get_int

# Returns digit at tenCounter's place in userCCNumber  
def getDigit(tenCounter, userCCNumber):
    return (userCCNumber % tenCounter) // (tenCounter * 0.1)

def main():

    userCCNumber = -1

    # Get credit card number from user  
    while userCCNumber < 0:
        userCCNumber = get_int("Your card number: ")
    
    # Keeps track of which digit in userCCNumber is being examined
    counter = 1

    # Used to isolate an integer value in userCCNumber
    tenCounter = 10

    # Sum of all examined digits in userCCNumber 
    grandSum = 0
    
    # Go through each digit in userCCNumber and add to grandSum depending on its position
    while tenCounter <= userCCNumber * 10:

        digit = getDigit(tenCounter, userCCNumber)

        # If counter is at an even numbered place
        if counter % 2 == 0:

            # Double the digit
            doubled = digit * 2
            
            remainder = doubled % 10

            # Is there now more than one digit in digit? 
            if doubled > 9:

               # Add both digits together  
               remainder = ((doubled - remainder) * 0.1) + remainder

            grandSum += remainder

        else:
            grandSum += digit

        counter += 1
        tenCounter *= 10

    # Counter overshot and needs to be rewound one
    counter -= 1

    # CCNumber can only be valid if grandSum is evenly divisible by 10
    if grandSum % 10 == 0:

        firstTwoDigits = int((getDigit(tenCounter * 0.1, userCCNumber) * 10)+(getDigit(tenCounter * 0.01, userCCNumber)))

        if firstTwoDigits == 34 or firstTwoDigits == 37:
            # Counter is equal to the number of digits in userCCNumber
            if counter == 15:
                print("AMEX")
                return
        elif int(firstTwoDigits * 0.1) == 4:
            if counter == 16 or counter == 13:
                print("VISA")
                return
        elif counter == 16:
            for i in range(51, 56):
                if i == firstTwoDigits:
                    print("MASTERCARD")
                    return
            
    print("INVALID")
    return

if __name__ == "__main__":
    main()